src/main/java/org/commonjava/util/partyline/impl/infinispan/model/InfinispanJFS.java

```diff
@@ -51,28 +55,39 @@ public LocalLockManager getLocalLockManager()
         return null;
     }
 
-    FileBlock getNextBlock( final FileBlock prevBlock ) throws InterruptedException
+    FileBlock getNextBlock( final FileBlock prevBlock, final FileMeta metadata) throws IOException
     {
-        UUID next = prevBlock.getNextBlockID();
-        FileBlock nextBlock = blockCache.get( next );
-        if ( nextBlock == null )
+        synchronized ( InfinispanJFS.this )
```

Comment:
We need to be careful with this, since it will lock all file access regardless of path. If it's really necessary to do that (if there's no better way to confine the lock to a path...and you can't use synchronized for that without causing memory bloat!), then we need to shrink the scope of this synchronized block as much as possible.

Having a while loop inside of this makes me a bit uneasy. Having a ReentrantOperation inside this makes me think that ReentrantOperation isn't actually locking anything, effectively...it's already locked at the provider instance level for all paths.

Looking through the rest of this commit, I'm not sure why this synchronized block is needed. Can you explain a bit more?

Reply:
This is resolved, see next comment in this file.